feat: new slider hook and container

jzempel opened the pull request:
## Description

This container encapsulates the [WAI-ARIA multi-thumb slider](https://www.w3.org/WAI/ARIA/apg/patterns/slidertwothumb/) and ports core logic directly from Garden's [MultiThumbRange](https://github.com/zendeskgarden/react-components/blob/main/packages/forms/src/elements/MultiThumbRange.tsx) component. Portions of the prop getter logic have been simplified by taking advantage of native event logic rather than relying on React state.

## Detail

The component was `yarn pack`'d to a `react-components` [branch](https://github.com/zendeskgarden/react-components/tree/jzempel/use-slider) where the existing `MultiThumbRange` has been drastically simplified and the entire test suite continues to pass.

## Checklist

Note that the PR is in _draft_ while testing remains a TBD by @jgorfine-zendesk.

- [x] :globe_with_meridians: Storybook demo is up-to-date (`yarn start`)
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [ ] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

jgorfine-zendesk commented on packages/slider/demo/stories/SliderStory.tsx:
```diff
@@ -0,0 +1,210 @@
+} from '@zendeskgarden/container-slider';
+import classNames from 'classnames';
+
+interface IComponentProps extends IUseSliderReturnValue {
+  max: IUseSliderProps['max'];
+  min: IUseSliderProps['min'];
+  minThumbRef: IUseSliderProps<Element, HTMLDivElement>['minThumbRef'];
+  maxThumbRef: IUseSliderProps<Element, HTMLDivElement>['minThumbRef'];
+  disabled: IUseSliderProps['disabled'];
+  rtl: IUseSliderProps['rtl'];
+}
```
Is there a particular reason this interface lives here, instead of in `types.ts`? (Is the story the only place we use it?)

jzempel replied:
Yes, this is local to the story and consistent with the demo typing for all other `react-containers` stories.

jgorfine-zendesk commented on packages/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,292 @@
+
+    const win = doc.defaultView || window;
+
+    win.addEventListener('resize', handleResize);
```
Would it be more performant, potentially, to listen to a change in the track size itself, rather than the whole window?

I'm thinking of a scenario in which the window resizes, but the track dimensions stay the same. In that scenario, we potentially would be firing this event unnecessarily.

jzempel replied:
I think there is that potential. However, the guideline for this PR was to port existing functionality over from the react-components `MultiThumbRange`. Garden may evaluate the use of `ResizeObserver` etc. as a future consideration across the code base. For now, we do not know of any performance or resizing concerns with the `MultiThumbRange` which has been in production for years.

jgorfine-zendesk commented on packages/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,292 @@
+   */
+
+  const getTrackProps = useCallback<IUseSliderReturnValue['getTrackProps']>(
+    ({ onMouseDown, ...other } = {}) => {
```
I've noticed that, when spreading objects, we use `...other` in some places and `...props` elsewhere. Just checking/confirming that the preference for this convention is to use `...other`?

jzempel replied:
We try to remain consistent with local codebase conventions – some of which have evolved more than have been documented. In `react-containers` that tends to be `...other`. As in: "the entire parameter is `props`, while the deconstructed one is `x` + `y` + `...other`".

smritimadan replied:
@jgorfine-zendesk Curious about the importance of this convention, so we don't end up overloading our contribution docs.

(Note though, that this shouldn't block this PR.)

jgorfine-zendesk replied:
@smritimadan totally agree that documenting this convention shouldn't block the PR. 👍 

I think it's important to do at some point, though. After all, if it's an expectation we have for the code base (and for contributors), then we should communicate it somewhere.

jgorfine-zendesk commented on packages/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,292 @@
+
+  const getThumbProps = useCallback(
+    (thumb: 'min' | 'max'): IUseSliderReturnValue['getMinThumbProps'] =>
+      ({ onKeyDown, onMouseDown, ...other }) => {
```
Same question here RE `...other` vs. `...props`?

jzempel replied:
ditto. https://github.com/zendeskgarden/react-containers/pull/483#discussion_r984579101

jgorfine-zendesk commented on packages/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,292 @@
+  );
+
+  const getThumbProps = useCallback(
+    (thumb: 'min' | 'max'): IUseSliderReturnValue['getMinThumbProps'] =>
```
Since this getter is used for both thumbs, it's a little confusing to see the return type reference only one of the thumbs (`getMinThumbProps`). Is there an opportunity here to create one centralized type? Or to change the naming here a little so it's representative of both thumbs?

jgorfine-zendesk commented on packages/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,292 @@
+          'aria-valuemax': thumb === 'min' ? position.maxValue : max,
+          'aria-valuenow': thumb === 'min' ? position.minValue : position.maxValue,
+          onKeyDown: composeEventHandlers(onKeyDown, handleKeyDown),
+          onMouseDown: composeEventHandlers(onMouseDown, handleMouseDown),
```
Is it worth attaching these event listeners to touch events, too? (In case a user is interacting with the range on a touch screen.)

jzempel replied:
I think that would be a worthwhile enhancement to the code ported over from `MultiThumbRange`. I'll take a closer look :eyes:.

jgorfine-zendesk reviewed (changes requested):
I have a couple of questions/comments sprinkled throughout, but I'm mostly requesting changes for two reasons:

1. I noticed one place where the word `Splitter` popped into the code.
2. I'm observing some odd behavior when using my mouse to click + drag the slider in both the Controlled and Uncontrolled demos. See details below. 👇 

## Unexpected click + drag behavior

### Uncontrolled: Max thumb not reaching slider max

Here's a screen recording of the odd behavior I observed when clicking + dragging the max thumb in the Uncontrolled story:

https://user-images.githubusercontent.com/93289772/193136382-2a2f3eea-a297-4950-8028-f08b94bb6ab2.mov

I wasn't able to get all the way back up to 100 until I released my click and re-initiated the drag. 

### Controlled: Min thumb skipping over value

Here's a screen recording of the odd behavior I observed when clicking + dragging the min thumb in the Controlled story:

https://user-images.githubusercontent.com/93289772/193137009-7a744606-e608-4fb5-9628-7744aae1a0f9.mov

I was able to get all the way down to 0; however, when moving my mouse back up again, the slider jumped from 0 to 2, for some reason.

jzempel commented on packages/slider/src/types.ts:
```diff
@@ -0,0 +1,86 @@
+  /** Defines the stepping interval */
+  step?: number;
+  /** Defines the jumping interval for keyboard page up/down navigation. Defaults to `step`. */
+  jump?: number;
```
WAI-ARIA support for `page up/down' as described by https://www.w3.org/WAI/ARIA/apg/example-index/slider/slider-multithumb.html#kbd_label. Allows a non-mouse user to more rapidly traverse before using the finer-grained `step` interval.

jzempel commented:
@jgorfine-zendesk re:

> I'm observing some odd behavior when using my mouse to click + drag the slider in both the Controlled and Uncontrolled demos.

I updated `setThumbPosition` to set state regardless of value changes. This improves the case for the uncontrolled slider which was not allowing a quick drag to settle out.

In addition I added checks for `isNaN`. We know that controlled components are less performant than uncontrolled in Storybook due to the overhead of [useArgs](https://storybook.js.org/docs/react/addons/addons-api#useargs) which keeps controls in sync. I was seeing issues with one control settling out change state would result in `NaN` calculations when the other control was manipulated. The code now handles this case which is much less likely to occur in production outside of Storybook.

jgorfine-zendesk commented on packages/slider/demo/stories/SliderStory.tsx:
```diff
@@ -0,0 +1,210 @@
+      getTrackProps,
+      getMinThumbProps,
+      getMaxThumbProps,
+      min = 0,
+      max = 100,
```
Is there a reason we have to set these values here, even though we're doing this by default in the hook? (Is it for Storybook purposes?)

jzempel replied:
Yes. Garden's standard is to externalize `defaultProps` values in Storybook controls for consumer clarity and to leave the others `undefined`.

jgorfine-zendesk commented on packages/slider/demo/stories/SliderStory.tsx:
```diff
@@ -0,0 +1,210 @@
+    }
+  };
+
+  return <Splitter />;
```
Confirmed changed. Thanks!

jgorfine-zendesk commented on packages/slider/demo/stories/SliderStory.tsx:
```diff
@@ -0,0 +1,210 @@
+}
+
+export const SliderStory: Story<IArgs> = ({ as, ...args }) => {
+  const Splitter = () => {
```
Confirmed changed. Thanks!

jgorfine-zendesk reviewed (changes requested):
Overall, this is looking great. The changes you made to the sliding functionality seem to have done the trick. Here's what I'm seeing on staging (on Chrome) now:

_Uncontrolled_

https://user-images.githubusercontent.com/93289772/193308198-04b923c6-af76-4014-9689-09e37b221a52.mov

_Controlled_

https://user-images.githubusercontent.com/93289772/193308311-bd348e1e-1744-4b9e-a215-153788c0538e.mov

_VoiceOver x Safari (LTR keyboard navigation)_

https://user-images.githubusercontent.com/93289772/193309833-69ab8ac1-dc61-429d-93b2-66e0a3165cb9.mov

However, when testing out the keyboard functionality in RTL mode, something went wrong:

https://user-images.githubusercontent.com/93289772/193309758-d21372ef-c09e-4619-ac97-dafd8e114f84.mov

The thumbs incremented and decremented by one, and then stopped. 🤔

jzempel commented on packages/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,317 @@
+        'data-garden-container-id': 'containers.slider.track',
+        'data-garden-container-version': PACKAGE_VERSION,
+        'aria-disabled': disabled,
+        onMouseDown: composeEventHandlers(onMouseDown, handleMouseDown),
```
The `onMouseDown` translates directly for touch.

jgorfine-zendesk replied:
Excellent.

jzempel commented:
@jgorfine-zendesk 

> However, when testing out the keyboard functionality in RTL mode, something went wrong[...] However, when testing out the keyboard functionality in RTL mode, something went wrong

Great find. I'll dig in next week and see what's up. :pray:

jgorfine-zendesk commented:
@jzempel @smritimadan thought for the future: should we create a ticket in the backlog to add support for `aria-valuetext`?

If you look at the [WAI-ARIA Roles, States, and Properties](https://www.w3.org/WAI/ARIA/apg/patterns/slidertwothumb/#wai-aria-roles-states-and-properties-18) section in the APG's Slider (Multi-Thumb) pattern, we aren't meeting this requirement, currently:
> If a value of aria-valuenow is not user-friendly, e.g., the day of the week is represented by a number, the [aria-valuetext](https://w3c.github.io/aria/#aria-valuetext) property is set to a string that makes the slider value understandable, e.g., "Monday".

jgorfine-zendesk commented:
> The good news is that we already have the desired support for this here via prop spread on the getter functions. 

Ah! @jzempel — good point. Is this where pattern documentation would come in, then? Like we just published for the Rich-text editor.

> The question is whether we should enhance `MultiThumbRange` in `react-components` to override value text.

I'm not quite sure what you mean by this. Mind telling me a bit more?

jgorfine-zendesk reviewed (approved):
Confirmed that the Slider is behaving nicely in the following interactions:
- Drag n' slide, Uncontrolled & Controlled
- Keyboard sliding with the arrow keys, Uncontrolled & Controlled, LTR & RTL

Nicely done!

I don't have a keyboard with Page Up, Page Down, Home, or End, so I haven't tested any of those. 😅 At least, not yet. (I have the ability to reprogram one of my keyboards so that it has those keys on it...but that'll take a bit.)

I'm approving this knowing that someone else will have to re-approve once I'm done writing the tests. 👍

jgorfine-zendesk commented:
Hey @jzempel — I also just noticed that we aren't rendering the `dir` attribute on any of the elements, when it's in a RTL layout. Is this on purpose, too?

jzempel commented:
@jgorfine-zendesk re:

> I'm noticing that we're not rendering the aria-disabled attribute on the thumbs, when they're disabled. Is this on purpose?

Yes. From https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-disabled:

> The state of being disabled applies to the element with aria-disabled="true" and all of its focusable descendants.

jzempel commented:
@jgorfine-zendesk re:

> I also just noticed that we aren't rendering the dir attribute on any of the elements, when it's in a RTL layout. Is this on purpose, too?

Yes. We only deal with the `rtl` prop in `react-containers`. That gives us the freedom to determine how to apply RTL styling. Using `dir` as has been done with this PR's storybook demo. Or using `direction` CSS styling as has been done with the `MultiThumbRange` component.

jzempel commented:
> I don't have a keyboard with Page Up, Page Down, Home, or End, so I haven't tested any of those.

You might try `Fn + Up/Down Arrow` on MacOS. https://apple.stackexchange.com/questions/26930/how-can-i-achieve-page-up-and-page-down-in-os-x

jgorfine-zendesk commented:
Hey @jzempel — adding touch interaction tests in now. I'm noticing that we don't have touch functionality on the Track right now, i.e., it only responds to mouse down to move the thumbs. Is this something we want to add in (plus a test for it)?

jzempel commented:
@jgorfine-zendesk shouldnt be needed as mousedown handling translates to a touch tap

jgorfine-zendesk commented:
> @jgorfine-zendesk shouldnt be needed as mousedown handling translates to a touch tap

Sounds good @jzempel!

Since this is coming up, I'm looking more into the different kinds of events: touch events vs. mouse events. vs. pointer events. Have you guys ever looked into using or switching to [Pointer Events](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events)? They've been [supported in React](https://reactjs.org/blog/2018/05/23/react-v-16-4.html) since 16.4. I ask because I wonder if they're more inclusive of different types of interactions/devices, and are potentially more accessible. 🤔 

(Note: I'm definitely not asking this to hinder the PR, but, rather, to talk/think about strategy.)

Francois-Esquire commented on packages/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,317 @@
+  rtl,
+  environment
+}: IUseSliderProps<T, M>): IUseSliderReturnValue {
+  const doc = environment || document;
```
With `environment` being `undefined`, this would throw a reference error in a non-browser environment. The same would apply for line 73 with `window`. This can be a possible followup item as we think through this problem.
![Screen Shot 2022-10-07 at 3 07 55 PM](https://user-images.githubusercontent.com/12474067/194636853-d93d4de4-3fc1-4cae-b071-add7d564cc66.png)

jzempel replied:
Yeah, this is why `environment` needs to be defined for SSR usage. We can consider more elaborate handling, but I think it is fair to say that SSR needs to give Garden something (bare-bones) to go on. I agree we can consider as a possible follow-up if/when SSR hits this container.

Francois-Esquire commented on packages/slider/src/useSlider.ts:
```diff
@@ -0,0 +1,317 @@
+
+    win.addEventListener('resize', handleResize);
+
+    return () => {
```
We should cancel any inbound debounce calls at this point.
```suggestion
    return () => {
      handleResize.cancel();
```
